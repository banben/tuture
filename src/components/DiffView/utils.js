function computeOldLineNumber(change) {
  if (change.isInsert) {
    return -1;
  }

  return change.isNormal ? change.oldLineNumber : change.lineNumber;
}

function computeNewLineNumber(change) {
  if (change.isDelete) {
    return -1;
  }

  return change.isNormal ? change.newLineNumber : change.lineNumber;
}

function getChangeKey(change) {
  if (!change) {
    throw new Error('change is not provided');
  }

  const {
    isNormal,
    isInsert,
    lineNumber,
    oldLineNumber,
  } = change;

  if (isNormal) {
    return 'N' + oldLineNumber;
  }

  const prefix = isInsert ? 'I' : 'D';
  return prefix + lineNumber;
}

export {
  computeOldLineNumber,
  computeNewLineNumber,
  getChangeKey,
}
