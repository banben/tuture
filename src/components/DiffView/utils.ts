import { Change as ChangeType } from '../../types';

function computeOldLineNumber(change: ChangeType): number {
  if (change.isInsert) {
    return -1;
  }

  return change.isNormal ? change.oldLineNumber : change.lineNumber;
}

function computeNewLineNumber(change: ChangeType): number {
  if (change.isDelete) {
    return -1;
  }

  return change.isNormal ? change.newLineNumber : change.lineNumber;
}

function getChangeKey(change: ChangeType): string {
  if (!change) {
    throw new Error('change is not provided');
  }

  const {
    isNormal,
    isInsert,
    lineNumber,
    oldLineNumber,
  } = change;

  if (isNormal) {
    return 'N' + oldLineNumber;
  }

  const prefix = isInsert ? 'I' : 'D';
  return prefix + lineNumber;
}

export {
  computeOldLineNumber,
  computeNewLineNumber,
  getChangeKey,
}
