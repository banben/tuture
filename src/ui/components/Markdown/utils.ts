const spliceStr = (str = '', typeStr: string, start: number, end: number) => {
  return str
    .slice(0, start)
    .concat(
      typeStr,
      str.slice(start, end),
      typeStr,
      str.slice(end, str.length),
    );
};

const insertStr = (str = '', typeStr: string, insertPosition: number) => {
  return str
    .slice(0, insertPosition)
    .concat(typeStr, str.slice(insertPosition, str.length));
};

export { spliceStr, insertStr };
