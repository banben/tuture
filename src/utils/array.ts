const isArray = (arr: any): boolean => {
  return Array.isArray(arr);
};

export {
  isArray,
};
