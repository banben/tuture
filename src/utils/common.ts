function handleAnchor(origin: string) {
  return origin.toLowerCase().replace(/ /g, '-');
}

function isClientOrServer() {
  return typeof window !== 'undefined' && window.document ? 'client' : 'server';
}

const reorder = (list: any[], startIndex: number, endIndex: number): any[] => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);

  return result;
};

const vwDesign = 1680;
const vwFontsize = 168;

function rem(px: number) {
  return px / vwFontsize;
}

export { handleAnchor, isClientOrServer, reorder, rem, vwDesign, vwFontsize };
