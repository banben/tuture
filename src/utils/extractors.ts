const extractCommits = (tuture: Tuture) => {
  const commits = tuture.steps.map((item: Step) => ({
    name: item.name,
    commit: item.commit,
  }));

  return commits;
};

const extractMetaData = (tuture: Tuture) => {
  const { name, topics, description } = tuture;
  const metadata = {
    name,
    topics,
    description,
  };

  return metadata;
};

export { extractCommits, extractMetaData };
