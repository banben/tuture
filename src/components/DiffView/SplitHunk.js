import React, { PureComponent } from 'react';
import classnames from 'classnames';

import SplitChange from './SplitChange';

import { getChangeKey } from './utils';

export default class SplitHunk extends PureComponent {
  keyForPair = (x, y) => {
    const keyForX = x ? getChangeKey(x) : '00';
    const keyForY = y ? getChangeKey(y) : '00';

    return keyForX + keyForY;
  };

  groupElements = (changes) => {
    let elements = [];

    for (let i = 0; i < changes.length; i++) {
      const current = changes[i];

      if (current.isNormal) {
        elements.push(
          ['change', this.keyForPair(current, current), current, current]
        );
      } else if (current.isDelete) {
        const next = changes[i + 1];

        // If an insert change is following a delete change
        // they should be displayed side by side
        if (next && next.isInsert) {
          i = i + 1;
          elements.push(
            ['change', this.keyForPair(current, next), current, next]
          );
        } else {
          elements.push(
            ['change', this.keyForPair(current, null), current, null]
          );
        }
      } else {
        elements.push(
          ['change', this.keyForPair(null, current), null, current]
        );
      }
    }

    return elements;
  }

  renderRow = ([type, key, oldValue, newValue], i) => {
    if (type === 'change') {
      return (
        <SplitChange
          key={`change${key}`}
          oldChange={oldValue}
          newChange={newValue}
        />
      );
    }
  }

  render() {
    const { hunk } = this.props;
    const elements = this.groupElements(hunk.changes);

    return (
      <tbody className={classnames('diff-hunk')}>
        {
          elements.map((element, i) => this.renderRow(element, i))
        }
      </tbody>
    );
  }
}
