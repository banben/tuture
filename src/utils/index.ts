import * as obj from './object';
import * as arr from './array';

const tutureUtilities = {
  ...arr,
  ...obj,
};

export default tutureUtilities;
