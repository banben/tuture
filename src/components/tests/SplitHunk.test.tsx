import React from 'react';
import renderer from 'react-test-renderer';

import SplitHunk from '../DiffView/SplitHunk';
import { hunk } from './utils/hunkData';

test('SplitHunk render successfully', () => {

  const component = renderer.create(
    <SplitHunk
      hunk={hunk}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
