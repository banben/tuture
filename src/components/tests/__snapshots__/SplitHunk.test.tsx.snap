// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitHunk render successfully 1`] = `
<tbody
  className="diff-hunk"
>
  <tr
    className="diff-line"
  >
    <td
      className="diff-gutter diff-gutter-omit"
    />
    <td
      className="diff-code diff-code-omit"
    />
    <td
      className="diff-gutter diff-gutter-insert"
      data-line-number={1}
    />
    <td
      className="diff-code diff-code-insert"
    >
      I am the coolest one
    </td>
  </tr>
  <tr
    className="diff-line"
  >
    <td
      className="diff-gutter diff-gutter-omit"
    />
    <td
      className="diff-code diff-code-omit"
    />
    <td
      className="diff-gutter diff-gutter-insert"
      data-line-number={2}
    />
    <td
      className="diff-code diff-code-insert"
    >
      I am the coolest one
    </td>
  </tr>
</tbody>
`;
